python_modules/dagster/dagster/_core/definitions/freshness_policy_sensor_definition.py

```diff
@@ -0,0 +1,381 @@
+        )
+
+    def current_minutes_late_by_key(self) -> Mapping[AssetKey, Optional[float]]:
+        """Returns a mapping from each monitored AssetKey with a FreshnessPolicy defined to a number
+        of minutes by which this asset is out of date with respect to its policy.
+        """
+        from dagster._core.execution.calculate_data_time import (
+            get_upstream_materialization_times_for_key,
+        )
+
+        current_time = pendulum.now()
+
+        statuses = {}
+        for asset_key, assets_def in self._assets_by_key.items():
+            if assets_def is None:
+                continue
+            freshness_policy = assets_def.freshness_policies_by_key.get(asset_key)
+            if freshness_policy is None:
+                continue
+            root_data_ids_and_timestamps = get_upstream_materialization_times_for_key(
+                instance=self.instance,
+                asset_key=asset_key,
+                upstream_asset_key_mapping=self._upstream_mapping,
+            )
+            statuses[asset_key] = freshness_policy.minutes_late(
+                evaluation_time=current_time,
+                upstream_materialization_times={
+                    AssetKey.from_user_string(k): v[1]
+                    for k, v in root_data_ids_and_timestamps.items()
+                },
+            )
+
+        # update cursor with new information
+        self.update_cursor(json.dumps({k.to_user_string(): v for k, v in statuses.items()}))
+        return statuses
+
+    def changed_freshness_statuses_by_key(self) -> Mapping[AssetKey, Optional[float]]:
+        """Returns a mapping from each monitored AssetKey with a FreshnessPolicy defined to a number
+        of minutes late.
+
+        This mapping will contain keys for which the freshness status has changed (i.e. started or
+        stopped being out of date) since the last tick on which statuses were evaluated.
+        """
+        return {
+            key: minutes_late
+            for key, minutes_late in self.current_minutes_late_by_key().items()
+            if (self._previous_statuses_by_key.get(key) == 0) != (minutes_late == 0)
+        }
+
+    def failing_freshness_statuses_by_key(self) -> Mapping[AssetKey, Optional[float]]:
+        """Returns a mapping from each monitored AssetKey with a FreshnessPolicy defined to a number
+        of minutes late.
+
+        This mapping will contain keys for which the data is currently out of date with respect to
+        their policy.
+        """
+        return {
+            key: minutes_late
+            for key, minutes_late in self.current_minutes_late_by_key().items()
+            if minutes_late != 0
+        }
+
+
+@experimental
+class FreshnessPolicySensorDefinition(SensorDefinition):
+    """Define an asset sensor that takes some action based on the SLA status of selected assets.
+
+    Users should not instantiate this object directly. To construct a
+    `FreshnessPolicySensor`, use :py:func:`dagster.
```

Comment:
Nitpick: `FreshnessPolicySensor` should either be "freshness policy sensor" or `FreshnessPolicySensorDefinition`.